fix(platform-browser): Make different SafeValue types incompatible values in TypeScript

mprobst opened the pull request:
Previously, because all `SafeValue`s were empty marker interfaces, users could
accidentally assign safe values into each other:

```
let myUrl: SafeUrl = ...;
let myHtml: SafeHtml = myUrl;
```

While this would fail at runtime (safety is enforced based on the values,
independent of TypeScript types), this can lead to very confusing code that
seems to pass around values of one type, which actually is of another type.

This change adds a `readonly kind` field that makes sure such accidental
assignments are compile time errors.

vicb commented on modules/@angular/platform-browser/src/security/dom_sanitizer.ts:
```diff
@@ -214,17 +236,22 @@ abstract class SafeValueImpl implements SafeValue {
 }
 
 class SafeHtmlImpl extends SafeValueImpl implements SafeHtml {
+  kind: SafeValueKind.HTML = SafeValueKind.HTML;
```
`kind: SafeValueKind = SafeValueKind.HTML;`
or
`kind = SafeValueKind.HTML;`

vicb replied:
Didn't know this was possible !
Thanks for clarification.

vicb commented on modules/@angular/platform-browser/src/security/dom_sanitizer.ts:
```diff
@@ -14,48 +14,68 @@ import {sanitizeUrl} from './url_sanitizer';
+ * This has no impact on the runtime enforcement of the safety, which is enforced by actually
+ * inspecting the runtime values (as opposed to the TypeScript type system).
+ *
+ * @stable
```
is this exported API ? (only them have stability markers)

rjamet commented:
I'm confused by the relationship between SafeUrl and SafeResourceUrl : SafeResourceUrl are accepted in SafeUrl context at the unwrapping stage (https://github.com/mprobst/angular/blob/28a821a8cce3d1dc67463139a11a4aba2092586b/modules/%40angular/platform-browser/src/security/dom_sanitizer.ts#L188), but there's no class inheritance between them ?

I assumed there was, and that this was why `buildTrustedUrl():SafeUrl { return SafeResourceUrl }` flowing into a non-AoT template's iframe src (resource url context at runtime) made sense. If there's not and I could have typed this with SafeHtml instead of SafeUrl without runtime errors... ouch, thanks for fixing this :)

Is there compile-time type checking at the template level? If so, this PR breaks using SafeResourceUrls in SafeUrl sinks, right? Dropping that relationship entirely is acceptable, I don't think I've ever seen much relying on this. It might just make SRU->SU context downgrades breaking, but we historically haven't had too many of those.

mprobst commented:
@rjamet good catch, I had forgotten about the relationship between `SafeResourceUrl` and `SafeUrl`. That was implemented by just changing the instanceof check to check for either or. I generally avoid inheritance if it's not needed, but in this case it would indeed have caught my oversight (requiring exactly `URL` instead of `URL|RESOURCE_URL` for `SafeUrl`).

Re type checking at the template level, we don't have that – after all you can pass in an arbitrary value to `sanitizer.sanitize(SecurityContext.HTML, ...)`, and it's the responsibility of the sanitizer to decide how to handle the value. In general, we might have values that can be sanitized at runtime or not (including resource URLs, e.g. imagine a whitelist in the sanitizer!), so whether a value passed in is safe cannot be enforced statically.

PTAL, I've updated the change list to make sure `SafeResourceUrl` is a subtype of `SafeUrl`, and actually changed kind to just be the `SecurityContext` enum.

rjamet commented:
LGTM, with the caveat that I don't really understand why this works :) This also looks like the kind of things where tests are useful, but it's lower priority.

Right for the type checking, I didn't think of that.

mprobst commented:
@rjamet tests would be nice, but we don't really have anything to test for "compile fails due to wrong types"...

mprobst commented:
Blocked on Angular going with TypeScript 2.0, which is schedule for Angular 3.0, which we'll start developing on in one month.

mprobst commented:
@IgorMinar please take a look, this changes the integration test to use TypeScript 2.0.

IgorMinar commented:
This will have to wait until after we branch off 2.3.x next Wednesday.

On Fri, Dec 2, 2016, 12:03 AM Martin Probst <notifications@github.com>
wrote:

> @IgorMinar <https://github.com/IgorMinar> please take a look, this
> changes the integration test to use TypeScript 2.0.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/12310#issuecomment-264322967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6O9TOvQAF6IFss4alTvwC9wxSoV1ks5rD1JAgaJpZM4KXOXO>
> .
>

mprobst commented:
@igorminar: is this unblocked now?

ocombe commented on packages/platform-browser/test/security/dom_sanitizer_spec.ts:
```diff
@@ -8,15 +8,14 @@
       const svc = new DomSanitizerImpl(null);
       const resourceUrl = svc.bypassSecurityTrustResourceUrl('http://hello/world');
-      t.expect(svc.sanitize(SecurityContext.URL, resourceUrl)).toBe('http://hello/world');
+      t.expect(svc.sanitize(SecurityContext.RESOURCE_URL, resourceUrl)).toBe('http://hello/world');
```
@mprobst I had to change this test to make it pass because of your change in dom_sanitizer.ts:173 `if (value instanceof SafeResourceUrlImpl || value instanceof SafeUrlImpl) {` --> `if (value instanceof SafeUrlImpl) {`
Let me know if that's ok, or if we shouldn't change this test

mprobst replied:
As replied above, the test here is correct, the change in the code is wrong. Please revert the code and keep the test as is (see also the test `it` message – this is intentional here).

mary-poppins commented:
You can preview 5cd4518 at https://pr12310-5cd4518.ngbuilds.io/.

mprobst commented:
@googlebot I consent to @ocombe's use of my original code.

mprobst commented on packages/platform-browser/src/security/dom_sanitizer.ts:
```diff
@@ -170,7 +171,7 @@ export class DomSanitizerImpl extends DomSanitizer {
         this.checkNotSafeValue(value, 'Script');
         throw new Error('unsafe value used in a script context');
       case SecurityContext.URL:
-        if (value instanceof SafeResourceUrlImpl || value instanceof SafeUrlImpl) {
```
This is an incorrect change, please revert. Probably an accidental mis-merge – the correct version accepts both `SafeResourceUrlImpl` and `SafeUrlImpl`.

ocombe replied:
Thanks, I knew that there was something fishy there!

mary-poppins commented:
You can preview 552c4c6 at https://pr12310-552c4c6.ngbuilds.io/.

mary-poppins commented:
You can preview d2945fc at https://pr12310-d2945fc.ngbuilds.io/.

mary-poppins commented:
You can preview dfb310c at https://pr12310-dfb310c.ngbuilds.io/.

IgorMinar commented:
Since this is changing the public api, we are waiting for a g3 global presubmit on this one to understand the impact on apps. Based on that we can decide if we should squeeze this into v5 or if it has to wait untill v6.

IgorMinar commented:
cl that needs love: cl/171304746 (I sent an email to Martin)

alexeagle commented:
Closing pull requests from 2016, sorry this got stale and was never merged. Please open a new one if you still care about this change.

mprobst commented:
Actually we should merge this one if we can; it was just stalled because it's an API breaking change. Can we get this in for the next version?

ocombe committed:
fix(platform-browser): make different SafeValues types incompatible values in TS

Previously, because all `SafeValue`s were empty marker interfaces, users
could accidentally assign safe values into each other:

    let myUrl: SafeUrl = ...;
    let myHtml: SafeHtml = myUrl;

While this would fail at runtime (safety is enforced based on the
values, independent of TypeScript types), this can lead to very
confusing code that seems to pass around values of one type, which
actually is of another type.

This change adds a `readonly kind` field that makes sure such accidental
assignments are compile time errors.

ocombe commented:
I've rebased the PR and updated the code. I didn't have to make this change:
![image](https://user-images.githubusercontent.com/265378/45742148-9a9c8580-bbf9-11e8-98ed-372868035da3.png)

The file doesn't exist anymore, and the tests seem to pass without it anyway.

I've also added `abstract` to `kind: SecurityContext;` line 207 in `abstract class SafeValueImpl` otherwise typescript wasn't happy. Let me know if that change is ok.

mary-poppins commented:
You can preview b1c1cf4 at https://pr12310-b1c1cf4.ngbuilds.io/.

mprobst commented:
We're still looking into whether it's possible to sync this into g3, please stand by.

ocombe commented:
ok, let me just update the golden file

mary-poppins commented:
You can preview db931a1 at https://pr12310-db931a1.ngbuilds.io/.

alexeagle commented:
I'm running the google3 presubmit. @mprobst let me know if I'm duplicating work

alexeagle commented:
I see, broken because of the platform_browser_safe redirection in first-party. I guess you are already working on that, reach out to me or Bowen if you need help or to ditch the task

mprobst commented:
I reached out to @rjamet to get help with the change.

mprobst commented:
We're working on fixing up the internal repo, we'll see if we can make it.

Am Mi., 26. Sep. 2018 um 09:21 Uhr schrieb Olivier Combe <
notifications@github.com>:

> Any news? The end of beta for v7 is coming fast
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/12310#issuecomment-424611696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAcKNHyaq0LK_7cpuCkS-qup5yG_bsjks5ueyr_gaJpZM4KXOXO>
> .
>

ocombe commented on packages/platform-browser/test/security/dom_sanitization_service_spec.ts:
```diff
@@ -9,7 +9,7 @@
 import {SecurityContext} from '@angular/core';
```
done

ocombe commented:
v7 rc0 was released, I guess that we missed the window

mary-poppins commented:
You can preview 3778d58 at https://pr12310-3778d58.ngbuilds.io/.

IgorMinar commented:
@rjamet @mprobst could we get this in some time in the next moth or so? is google3 ready?

mary-poppins commented:
You can preview e5bb830 at https://pr12310-e5bb830.ngbuilds.io/.

mary-poppins commented:
You can preview b318021 at https://pr12310-b318021.ngbuilds.io/.

petebacondarwin commented on aio/src/app/custom-elements/toc/toc.component.spec.ts:
```diff
@@ -30,6 +31,33 @@ describe('TocComponent', () => {
+      tocItem('H2 4', 'h2', 'fizz/buzz#h2-four', '<i>H2 <b>four</b></i>'),
+    ];
+  }
+
```
Why move this block to the top?

mprobst replied:
It needs access to `tocItem`, which in turn needs access to `sanitizer`, so it has to be within the `describe` scope, instead of at the bottom.

petebacondarwin replied:
Yeah, it is just awkward, since there are now some helpers at the top and some at the bottom.

Instead, since it is `Injectable` you could make the `TestService` depend directly on the `DomSanitizer`, and if necessary make the `getTestTocList()` and `tocItem()` functions private methods on `TestService`... ? This would also save on having to add in a `let` to share  the `sanitizer` around the tests.

mprobst replied:
Several test methods use `tocItem` outside of the component (at least in the other file). I could change `tocItem` to explicitly take the `sanitizer` as an argument and pass it around, but I'm not sure that's better?

petebacondarwin commented on aio/src/app/custom-elements/toc/toc.component.spec.ts:
```diff
@@ -30,6 +31,33 @@ describe('TocComponent', () => {
   }
 
+  function tocItem(title: string, level = 'h2', href = '', content = title): TocItem {
+    return {title, href, level, content: sanitizer.bypassSecurityTrustHtml(content)};
```
+1

mary-poppins commented:
You can preview a6548d5 at https://pr12310-a6548d5.ngbuilds.io/.

petebacondarwin reviewed (approved):
LGTM from a `docs-infra` point of view - although I think that moving the test helper functions around causes unnecessary git-diff churn.

IgorMinar commented:
@mprobst is g3 ready for this change? if so could you rebase one more time plz and add re-request my review? thanks!

mprobst commented:
I don't think so, at least I think Raphael and I both didn't have time to
follow through with this.

mhevery commented:
Closing because PR is out of date, and TrustedTypes is the new preferred approach.
